org.eclipse.xtext.ui/src/org/eclipse/xtext/ui/wizard/template/TemplateProcessor.java

```diff
@@ -0,0 +1,166 @@
+			for (ClassDeclaration annotatedClass : annotatedSourceElements) {
+				actualPropertyContents = propertyContentMap.get(this.getMessagesProperties(annotatedClass));
+				doGenerateCode(annotatedClass, context);
+			}
+			saveFileMaps(annotatedSourceElements, context);
+		}
+	}
+
+	@Override
+	public void doGenerateCode(ClassDeclaration annotatedClass, @Extension CodeGenerationContext context) {
+		synchronized (TemplateProcessor.LOCK) {
+			String propertyContents = generatePropertiesFile(annotatedClass, context);
+			generateMessagesClass(propertyContents, annotatedClass, context);
+		}
+	}
+
+	// read the "label" and "description" from the ProjectTemplateAnnotation and update "messages.properties" with the values
+	private String generatePropertiesFile(ClassDeclaration annotatedClass, @Extension CodeGenerationContext context) {
+		String label = replaceNewlines(getLabel(annotatedClass, context));
+		String description = replaceNewlines(getDescription(annotatedClass, context));
+		String labelLineStart = annotatedClass.getSimpleName() + "_Label=";
+		String labelLine = labelLineStart + label;
+		if (actualPropertyContents.contains(labelLineStart)) {
+			actualPropertyContents = actualPropertyContents.replaceFirst("(?m)^" + labelLineStart + ".*$", labelLine);
+		} else {
+			actualPropertyContents = actualPropertyContents + (labelLine + System.lineSeparator());
+		}
+		String descriptionLineStart = annotatedClass.getSimpleName() + "_Description=";
+		String descriptionLine = descriptionLineStart + description;
+		if (actualPropertyContents.contains(descriptionLineStart)) {
+			actualPropertyContents = actualPropertyContents.replaceFirst("(?m)^" + descriptionLineStart + ".*$", descriptionLine);
+		} else {
+			actualPropertyContents = actualPropertyContents + (descriptionLine + System.lineSeparator());
+		}
+		propertyContentMap.put(getMessagesProperties(annotatedClass), actualPropertyContents);
+		return actualPropertyContents;
+	}
+
+	// generate a "Messages.java" out of the given content from "messages.properties"
+	private void generateMessagesClass(String propertyContents, ClassDeclaration annotatedClass, @Extension CodeGenerationContext context) {
+		Path classFile = annotatedClass.getCompilationUnit().getFilePath().getParent().append("Messages.java");
+		String contents = "";
+		if (context.exists(classFile)) {
+			String oldContents = context.getContents(classFile).toString();
+			if (oldContents.trim().startsWith("/*") && oldContents.contains("*/") && oldContents.length() > oldContents.indexOf("*/") + 2) {
+				contents = contents + (oldContents.substring(0, oldContents.indexOf("*/") + 2) + System.lineSeparator());
+			}
+		} else {
+			contents = contents + (annotatedClass.getCompilationUnit().getDocComment() + System.lineSeparator());
+		}
+		contents = contents + "package " + annotatedClass.getCompilationUnit().getPackageName() + ";" + System.lineSeparator();
+		contents = contents + 
+				System.lineSeparator() +
```

Comment:
TIL: CodeGenerationContext does not expose a line separator that should be used for a given URi.